### ISSUE TITLE
Implement a "rapid rebuild" mode that only rebuilds tex and not dependencies

### ISSUE THREAD

**Issue by dfm** [showyourwork/showyourwork#298]
From @adrn:

> Make it easy to just rebuild the tex when in rapid text/paper editing mode and don’t care about updating figures or datasets

In the ideal world, snakemake already does this: it'll only generate the figures/datasets if the scripts change. Perhaps @adrn is considering a case where the script _is_ changed, but you still want to ignore it?

I think that one way we could achieve something like this would be by conditionally applying the `ancient` flag to the inputs of the compile rule.

**Comment by adrn** [comment]
> a case where the script is changed, but you still want to ignore it?

Yes, exactly!

**Comment by kazewong** [comment]
Is it possible to have a turbo mode that skips all the installation and scripts? Sometimes I want to just change the text a bit but then it takes >10 seconds to compile the source.

**Comment by adrn** [comment]
`ancient()` flag is here: https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#ignoring-timestamps

Within draft mode, do we want to apply this more globally for output files so that even data files don't get re-generated if their outputs exist? A specific use case I can think of: I have a data file downloaded from a data archive from some past work or script (or maybe it's a subset of the data that I want to prototype with), and I write a rule to eventually generate that file in the workflow, but I want snakemake to just use the existing file without re-running a rule to re-generate that file. Currently, this can be done by putting `ancient()` around that specific output file, so I will add that to the FAQ. But maybe draft mode will want to apply this more broadly? Not sure.

**Comment by mathren** [comment]
I'm still working on this, and I'm finding a behavior I don't understand.

I have a draft mode setup that first runs the `preprocess.smk` rules to build the tree of dependencies. The idea is to then run a set of `draft` rules that:
- loops over the dependencies and if the file already exists copies it to /static/ otherwise generate a dummy one (figure/dataset)
- changes all the paths of the dependencies to be in `static`

My understanding is that after the rules in `preprocess.smk` the tree of dependencies should be defined, and indeed in my test repository I do have `src/.showyourwork/preprocess/showyourwork.xml` which lists the tex file dependencies produced by the last but one rule of `preprocess.smk`.

My idea was to then use the `config["dependencies"][config["ms_tex"]]`  (which is an input to the rule in `compile.smk` that "Setup the temporary files for compilation")  as the starting point to create missing dependencies and moving existing ones to /static.

However, when (after successful completion of the preprocess step), if I try to access that, I get a `KeyError` because of `[config["ms_tex"]]`, e.g. when trying to print this into a logfile:

```
logfile.writelines(config["dependencies"][config["ms_tex"]])
KeyError: 'src/tex/ms.tex'
```

and in general, it seems like `config["dependencies"]` is empty at this point, which is not what I expect after the successful execution of the `preprocess` set of rules. Does anyone have insight on what I'm not getting?

**Comment by dfm** [comment]
@mathren — Interesting! The `preprocess.smk` run will generate a file called `.showyourwork/config.json` which should include this info. Then in `build.smk` this file is loaded into the `config` variable:

https://github.com/showyourwork/showyourwork/blob/532b1bc66e73bd281d7333c9384c7823ac0515f7/showyourwork/workflow/build.smk#L27

In your workflow, is `.showyourwork/config.json` being created with the dependencies listed? And if so, are you then using that as your config file in `draft.smk`?

**Comment by mathren** [comment]
no, the `config.json` is not being created somehow, though the preprocess step does not raise any errors. 
From the last attempt, the logs show:
```
Preprocess: Setting up...
Preprocess: Generating XML tree...
warning: Tectonic does not support `config' special. Ignored.
Preprocess: Setting up the workflow...
Draft: fixing dependencies...
```
where the fact that after the "setting up the workflow" message I see the message "Draft: ..."  makes me think that the previous step completed successfully.

However, now that I am looking into this, my new  draft rule in development right now has `output: config[config_json]` (copied from `preprocess.smk`) which seems to be at least part of the problem. Removing that I do see the config.json being created.

**Comment by mathren** [comment]
Is there a good "dummy" output I can use to make sure the `draft` rule is triggered?
